### ISSUE TITLE
ERROR: connect ECONNREFUSED 127.0.0.1:27017

### ISSUE THREAD

**Issue by pinkworld123** [Tzahi12345/YoutubeDL-Material#614]
Hello, 

I try to run image from docker hub and i got this error .ERROR: connect ECONNREFUSED 127.0.0.1:27017

my docker-compose.yml

```
version: 3.7
services:
  ytdl_material:
    image: tzahi12345/youtubedl-material:nightly
    container_name: youtubedl-material
    restart: unless-stopped
    volumes:
      - ./data/appdata:/app/appdata
      - ./data/audio:/app/audio
      - ./data/video:/app/video
      - ./data/subscriptions:/app/subscriptions
      - ./data/users:/app/users
    environment:
      UID: 1006
      GID: 1006
      ALLOW_CONFIG_MUTATIONS: 'true'
    ports:
      - 17442:17442
    networks:
      - youtubedl-material   


```

**Comment by MattNer0** [comment]
Hello,
I was having the same problem so I checked the docker-compose.yml from the repo [here](https://github.com/Tzahi12345/YoutubeDL-Material/blob/master/docker-compose.yml) and noticed that the mongodb instance was moved to another container.

I guess you can try to play with the env `ytdl_mongodb_connection_string` or `ytdl_use_local_db`, but I didn't test it.
In my case I just recreated both containers using the updated `docker-compose.yml` from the repo.

Cheers

**Comment by GlassedSilver** [comment]
> Hello, I was having the same problem so I checked the docker-compose.yml from the repo [here](https://github.com/Tzahi12345/YoutubeDL-Material/blob/master/docker-compose.yml) and noticed that the mongodb instance was moved to another container.
> 
> I guess you can try to play with the env `ytdl_mongodb_connection_string` or `ytdl_use_local_db`, but I didn't test it. In my case I just recreated both containers using the updated `docker-compose.yml` from the repo.
> 
> Cheers

YTDL-M never included a MongoDB backend within its own container and therefore nothing got ever moved.

What we used before was a single `.json` file as our "local db". By contrast, MongoDB can be put anywhere, even on a different physical host - or in the case of a `docker-compose.yml` file: in a separate container.

When the variable to configure MongoDB got introduced, that was the time when we actually enabled to use it to begin with.

That being said, maybe your reply helps @pinkworld123 out, thank you for your feedback.

If this doesn't get resolved, please don't hesitate to comment further, we're absolutely happy to help! 👍

**Comment by pinkworld123** [comment]
Thanks for reply. 

@GlassedSilver I was debugging and found that by default tries to connect to mongodb and when its not found after few retries it uses json. 

So basically default image is working fine just need to wait few minutes.

**Comment by sunjam** [comment]
The [docker-compose.yml](https://raw.githubusercontent.com/Tzahi12345/YoutubeDL-Material/master/docker-compose.yml) does not work for me when attempting to use mongo.

```
getaddrinfo EAI_AGAIN ytdl-mongo-db
ECONNREFUSED 127.0.0.1:27017
Failed to connect to MongoDB, using Local DB as a fallback. Make sure your MongoDB instance is accessible, or set Local DB as a default through the config.
```

**Comment by GlassedSilver** [comment]
> Thanks for reply.
> 
> @GlassedSilver I was debugging and found that by default tries to connect to mongodb and when its not found after few retries it uses json.
> 
> So basically default image is working fine just need to wait few minutes.

Now that I think about it, we absolutely do advise to use MongoDB, because it’s MUCH MUCH more stable.

It resolved many headaches for me and others.

Are you still on the json path and want to move to Mongo?

**Comment by GlassedSilver** [comment]
> The [docker-compose.yml](https://raw.githubusercontent.com/Tzahi12345/YoutubeDL-Material/master/docker-compose.yml) does not work for me when attempting to use mongo.
> 
> ```
> getaddrinfo EAI_AGAIN ytdl-mongo-db
> ECONNREFUSED 127.0.0.1:27017
> Failed to connect to MongoDB, using Local DB as a fallback. Make sure your MongoDB instance is accessible, or set Local DB as a default through the config.
> ```
> 
>  ```
> version: "2"
> services:
>     ytdl_material:
>         environment: 
>             ALLOW_CONFIG_MUTATIONS: 'true'
>             ytdl_mongodb_connection_string: 'mongodb://ytdl-mongo-db:27017'
>             ytdl_use_local_db: 'false'
>             write_ytdl_config: 'true'
>         restart: always
>         depends_on:
>             - ytdl-mongo-db
>         volumes:
>             - ./appdata:/app/appdata
>             - ./audio:/app/audio
>             - ./video:/app/video
>             - ./subscriptions:/app/subscriptions
>             - ./users:/app/users
>         ports:
>             - "8998:17442"
>         image: tzahi12345/youtubedl-material:nightly
>     ytdl-mongo-db:
>         image: mongo
>         ports:
>             - "27017:27017"
>         logging:
>             driver: "none"          
>         container_name: mongo-db
>         restart: always
>         volumes:
>             - ./db/:/data/db
> ```

Found the issue.

Change MongoDB’s container name to `ytdl-mongo-db` so it matches the URL passed to YTDL-M container configs, so that it looks like this:

```
version: "2"
services:
    ytdl_material:
        environment: 
            ALLOW_CONFIG_MUTATIONS: 'true'
            ytdl_mongodb_connection_string: 'mongodb://ytdl-mongo-db:27017'
            ytdl_use_local_db: 'false'
            write_ytdl_config: 'true'
        restart: always
        depends_on:
            - ytdl-mongo-db
        volumes:
            - ./appdata:/app/appdata
            - ./audio:/app/audio
            - ./video:/app/video
            - ./subscriptions:/app/subscriptions
            - ./users:/app/users
        ports:
            - "8998:17442"
        image: tzahi12345/youtubedl-material:nightly
    ytdl-mongo-db:
        image: mongo
        ports:
            - "27017:27017"
        logging:
            driver: "none"          
        container_name: ytdl-mongo-db
        restart: always
        volumes:
            - ./db/:/data/db
```

I will make a PR to fix this inconsistency.

Sometimes it takes a while and some distance to spot these sorts of issues. (I personally deployed differently, so my working setup never challenged this docker-compose.yml.

**Comment by GlassedSilver** [comment]
I think we should also reconsider the decision to fallback to internal db on failing connection attempts.

**Comment by sunjam** [comment]
I thought it was the Service name being passed as opposed to the Container name.

Made the container name change to `ytdl-mongo-db` and the problem persists, even after removing and re-adding images/containers/volumes.
```
ERROR: connect ECONNREFUSED 127.0.0.1:27017
ERROR: getaddrinfo EAI_AGAIN ytdl-mongo-db
```

**Comment by GlassedSilver** [comment]
> I thought it was the Service name being passed as opposed to the Container name.
> 
> Made the container name change to `ytdl-mongo-db` and the problem persists, even after removing and re-building images/containers/volumes.
> 
> ```
> ERROR: connect ECONNREFUSED 127.0.0.1:27017
> ERROR: getaddrinfo EAI_AGAIN ytdl-mongo-db
> ```

Oh good catch, yeah I think it is the way you said it. I must be mistaken. I run most of my stuff without docker compose, that's why I mistook this, probably.

You may still try this:

https://docs.docker.com/compose/compose-file/compose-file-v2/#networks

If you force them into the same specifically defined network that may do the trick. Alternatively there's the legacy option to link the two containers, but I don't know whether using a legacy option is a good long-term approach:

https://docs.docker.com/compose/compose-file/compose-file-v2/#links

**Comment by stumpy3toes** [comment]
I had the same problem and turned on logging for the mongodb container. Turns out that in my case mongodb was throwing an error and refusing to start. This is because on the newer versions of mongo (v5+) they have a requirement for ARM builds where the host CPU must meet the ARM spec ARMv8.2-A. I guess raspberry pi 4 doesn't meet that or it can't detect that it can when in docker. Either way I just set the image to an older version:
```yaml
image: mongo:4.4.6
```

TL;DR mongodb version 5+ doesn't work on raspberry pi 4 in docker, so I set the mongodb image to v4.4.6

**Comment by GldRush98** [comment]
How do I get it to stop trying to connect to mongodb on startup and always just use the local json file?
I have tried setting:
ytdl_use_local_db: true
and
use_local_db: true
but neither one made it skip the mongodb check on start that makes it start up so slow.

In face, the docs say that use_local_db defaults to true... so this shouldn't even need to be set. It shouldn't be trying to connect to mongodb unless specifically told to, no?

**Comment by GlassedSilver** [comment]
@stumpy3toes The latest and greatest Raspberry Pi, model 4B uses a Broadcom BCM2711 SoC which is based on Cortex-A72.

[Checking this tablet](https://en.wikipedia.org/wiki/Comparison_of_ARMv8-A_processors), we can learn that in fact, the ARM spec for it is 8.0-A.

@GldRush98 I would like to relegate that question to @Tzahi12345 since I'm personally a bit less involved with the code itself. That being said, I hope you know that the json file config is error-prone and really just useful for quick test deployments, very small installs, especially when not using subscriptions [or not much at least] and if you don't mind metadata possibly getting lost and rebuilt.

You can check out this issue if you like, it's one of a few about database loss and points to mongodb as solving this. #425

However, of course the mongodb check should skip if it's disabled.

**Comment by Tzahi12345** [comment]
> How do I get it to stop trying to connect to mongodb on startup and always just use the local json file? I have tried setting: ytdl_use_local_db: true and use_local_db: true but neither one made it skip the mongodb check on start that makes it start up so slow.
> 
> In fact, the docs say that use_local_db defaults to true... so this shouldn't even need to be set. It shouldn't be trying to connect to mongodb unless specifically told to, no?

I think I fixed the issue in #657!

**Comment by GldRush98** [comment]
> I think I fixed the issue in #657! Commit: [55d4f74](https://github.com/Tzahi12345/YoutubeDL-Material/commit/55d4f746c352bf1cbe78c3e8725177b866bb6898)

Comfirmed! I patched my local app.js file and it started up instantly, zero delay. Fantastic!

**Comment by Tzahi12345** [comment]
Awesome great to hear, I think we're safe to close this as all the issues posted here have been resolved. If I'm wrong please correct me and I'll reopen!

**Comment by p6002** [comment]
Any way to fix it? Almost a year and i still cannot use official docker compose because mongo not working.
